tests/ui/lint/dead-code/unused-adt-impls-trait.rs

```diff
@@ -0,0 +1,34 @@
+#![deny(dead_code)]
+
+struct Used;
+struct Unused; //~ ERROR struct `Unused` is never constructed
+
+pub trait PubTrait {
+    fn foo(&self) -> Self;
+}
+
+impl PubTrait for Used {
+    fn foo(&self) -> Self { Used }
+}
+
+impl PubTrait for Unused {
+    fn foo(&self) -> Self { Unused }
+}
+
+trait PriTrait {
+    fn foo(&self) -> Self;
+}
+
+impl PriTrait for Used {
+    fn foo(&self) -> Self { Used }
+}
+
+impl PriTrait for Unused {
+    fn foo(&self) -> Self { Unused }
+}
+
+fn main() {
+    let t = Used;
```

Comment:
`Used` is used because this variable is constructed (at least).
If the `foo` methods didn't take `self`, then it would be more clear that `Used` is used due to trait impls.

Reply:
In this implementation, if the `foo` methods didn't take `self`, it will be marked alive when the `foo` is used. Because if it only return `Self`, we can only know the trait item `foo` is used and it may construct the `Unused`, then we can only assume `Unused` is used because we are not sure whether the impl item is not used.

Reply:
Then `let t2 = Unused;` should also be added, because right now the test is not actually testing the difference.

Reply:
I think not, if `let t2 = Unused` is added, then `Unused` is constructed. The difference in this test is the two traits are both used, but we will only lint the struct which is constructed. Without this change, we cannot lint `Unused`.